src/client.ts

```diff
@@ -405,51 +398,353 @@ export class UXDClient {
+        zoAccount: depository.zoAccountPda,
+        zoOpenOrders: openOrdersPda,
+        zoState: zo.state.pubkey,
+        zoStateSigner: zo.state.signer,
+        zoDexMarket: market.address,
+        zoControl: control,
+        zoDexProgram: zo.dexProgramId,
+        systemProgram: SystemProgram.programId,
+        zoProgram: zo.program.programId,
+        rent: SYSVAR_RENT_PUBKEY,
+      },
+      options: options,
+    });
+  }
+
+  public async createMintWithZoDepositoryInstruction(
+    collateralAmount: number, // UI
+    slippage: number,
+    controller: Controller,
+    depository: ZoDepository,
+    zo: Zo,
+    user: PublicKey,
+    options: ConfirmOptions,
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const userCollateralATA = await utils.token.associatedAddress({
+      mint: depository.collateralMint,
+      owner: user,
+    });
+    const userRedeemableATA = await utils.token.associatedAddress({
+      mint: controller.redeemableMintPda,
+      owner: user,
+    });
+    const limitPrice = depository.getLimitPrice(slippage, 'short', zo);
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
+    );
+
+    const limitPriceBn = market.priceNumberToLots(limitPrice);
+    const feeMultiplier = 1 - ZO_FUTURE_TAKER_FEE;
+    const maxBaseQtyBn = market.baseSizeNumberToLots(collateralAmount);
+    // Removing the odd lots
+    const maxQuoteQtyNativeBn = new BN(
+      Math.round(
+        limitPriceBn
+          .mul(maxBaseQtyBn)
+          .mul(market.decoded['quoteLotSize'])
+          .toNumber() * feeMultiplier
+      )
+    );
+    console.log('QUOTE LOT SIZE', market.decoded['quoteLotSize'].toNumber());
+    console.log('BASE LOT SIZE', market.decoded['baseLotSize'].toNumber());
+
+    console.log('collateralAmount', collateralAmount);
+    console.log('Limit price', limitPriceBn.toNumber());
+    console.log('maxQuoteQtyBn', maxQuoteQtyNativeBn.toNumber());
+    console.log('maxBaseQtyBn', maxBaseQtyBn.toNumber());
+    return this.instruction.mintWithZoDepository(
+      maxBaseQtyBn,
+      maxQuoteQtyNativeBn,
+      limitPriceBn,
+      {
+        accounts: {
+          user,
+          payer: payer ?? user,
+          controller: controller.pda,
+          depository: depository.pda,
+          redeemableMint: controller.redeemableMintPda,
+          userCollateral: userCollateralATA,
+          userRedeemable: userRedeemableATA,
+          zoAccount: depository.zoAccountPda,
+          zoState: zo.state.pubkey,
+          zoStateSigner: zo.state.signer,
+          zoCache: zo.state.cache.pubkey,
+          zoVault: zo.state.getVaultCollateralByMint(
+            depository.collateralMint
+          )[0],
+          zoControl: marginAccount.control.pubkey,
+          zoOpenOrders:
+            await depository.getZoAccountCollateralOpenOrdersAccountKey(
```

Comment:
avoid putting `await` of async operations in long-ass object declarations, prefer `await` in a variable and then use that later on